### DISCUSSION TITLE
ffpmeg hardware acceleration - Intel vs AMD

### DISCUSSION THREAD

**Discussion by shred86** [blakeblackshear/frigate#1566] [General]
I’m in the market for a new mini PC to replace my current one (AWOW AL34 w/ Intel Celeron N3450) that serves as my home server and I think I’ve narrowed it down to the Beelink GTI (Intel CPU & GPU) or GTR (AMD CPU & GPU). I’m planning on swapping over my M.2 Coral EdgeTPU but in terms of getting the best performance for ffmpeg hardware acceleration, which CPU & GPU combination would be better?

- Intel Core i5-8259U w/ Intel Iris plus 655 GPU
- AMD Ryzen 5 3550H w/ AMD Radeon Vega 8

The benchmarks I’ve seen show the Ryzen 3550H and Radeon Vega 8 as both being better performers overall (higher energy consumption too), but I wasn’t sure how that translates to ffmpeg performance for processing video streams. I run a lot of other containers that aren’t very resource intensive, but I’d just like to optimize the hardware for Frigate since that’s the most resource intensive service I’m running right now.

Additionally, if you have any other mini PC recommendations that have at least two M.2 slots, I’d love to hear about those as well. I’ve found a couple mini PCs with removable M.2 wifi cards but most of them sit immediately underneath the M.2 SSD. Doesn’t seem like a smart idea to run an EdgeTPU underneath an SSD, so I’m really looking for mini PCs that have M.2s that are not co-located.

**Comment by b3nj1** [comment]
Hi @shred86 , what did you go with, and how is it working out for you?
Thanks!

**Reply by shred86** [reply]
I ended up going with the [Odyssey x86](https://www.amazon.com/gp/product/B08FD3QHQC/ref=ppx_yo_dt_b_asin_title_o03_s00?ie=UTF8&psc=1) from Seeed Studios. The mini PCs I was looking at when I made the original post are overkill for what I needed. I run about 10 docker containers for stuff like Home Assistant, Pihole, Samba, Unifi controller, etc, so the Intel Celeron J4125 with 8GB RAM is plenty. Frigate consumes the most resources and even with everything, I’m sitting around 20-25% CPU usage and about 3.5GB of RAM being used.

Of the higher end mini PCs I was looking at, the Beelink GTI series seems to be one of the few with an M.2 layout that would work well with an Edge TPU. I’ve found other mini PCs will stack their M.2 slots because it’s usually for something like a wifi card that can sit next to an SSD, but I would not want an Edge TPU sitting that close to an SSD.

**Reply by ferbulous** [reply]
Thanks, some question:
1. Are you using low res for detection on all cameras added into frigate?
2.  Can you use dual edge tpu on the odyssey's m.2 slot, and are both detected?

**Reply by shred86** [reply]
1. I’m using 720p/15fps streams for detection (5 fps set in frigate for detect) on four cameras. The reason I keep it at 15fps on the camera side is because I use the same stream for RTMP as well. 

2. I believe so since it’s a 4x PCIe slot, but I personally haven’t tried. I’m using just a single Edge TPU. If you need a dual edge TPU though, I think you’d need a better CPU too because I’m assuming you’d be running a lot of camera to need more than one TPU. I’m seeing about 8-11ms inference times which should handle up to 100fps.

**Comment by b3nj1** [comment]
Thanks for all the info. It’s really helpful to know.

I have 12 cameras on a Raspberry Pi 4/4GB with USB Coral. With detect using
 ~VGA 5 fps from camera, it handles everything fine with a load in the 2-4
range (4 core). I want full res clips for all events and full res recording
for 5 cameras, but that is too much for it (load spikes to 20+ and lots of
drops). Full res being 4k at 15fps for these cameras. With frigate stopped,
I can use  ffmpeg to copy all RTSP streams at full res with a load of 0.5.
So, it seems like it should be possible to do the low res detect plus full
res clips/record without the system falling over like it does.  I’ve been
experimenting with ways to achieve that without spending money, but I might
have to get a more powerful server.

**Reply by shred86** [reply]
Yeah, 12 cameras seems like a lot for an RPi 4. I’m actually surprised it handles all 12 even at VGA resolutions! There’s a lot of mini PCs out there with the Intel Celeron J4125 that you can find for under $200, like the AWOW AK41 or NY41. They don’t have M.2 PCIe slots that work with an Edge TPU but you could just use your Coral USB. I’ve seen them go on sale for as low as $160. I had the AK41 for about three weeks and I noticed the CPU cooling wasn’t too good. With the setup I’m running now (20-25% CPU usage), I was seeing around 55-60 deg C and this is with an additional 120mm fan that was blowing air across the mini PC since I have it in a server rack. The Odyssey X86 with the same setup and fan sits around 45-dec C. The NY41 has a different heat sink setup and supposedly much better cooling.

I have four Loryta (Dahua) IPC-T5442TM-AS cameras which are 4MP. I’m running 24/7 recording and clips at 4MP/15fps no problem (with the detect and RTSP at 720p/15fps, but 5fps set in Frigate). My understanding is the recording and clips shouldn’t take up much CPU since it’s just directly recording the H.264 video feed to your hard drives. It’s just ffmpeg for decoding the detect streams that eats up CPU.

**Reply by blakeblackshear** [reply]
You really shouldn't see much CPU usage if any from using the 4k stream for the record role. Frigate literally just runs ffmpeg in the background to copy the rtsp streams to segments. You may want to open an issue with your config if you aren't seeing this.

**Reply by b3nj1** [reply]
Thanks. I agree. Since this morning, I've been running Frigate using low res for detects+clips, and outside Frigate, I've been saving full res at the same time. It's been running without issue at a load ranging from 2-6. When I get some time to upgrade to 0.9.x, I'll give it another try to put everything within Frigate and report back my findings.
